accepted/cranelift-isel-pre-rfc.md

```diff
@@ -0,0 +1,798 @@
+
+Minimal coupling also allows for use of the lowering rules in
+different contexts: for example, it would likely help with
+verification efforts (below), and makes testing/fuzzing much more
+practical with a smaller testing harness.
+
+## Requirement 5: Ability to Encode Lowering Invariants
+
+While lowering code, there are a number of rules that one must
+carefully observe in order to maintain correctness. A simple program
+composed only of pure arithmetic operators poses little worry: one can
+always duplicate an operator, or move it elsewhere in the program, as
+long as the dataflow remains intact. In contrast, (i) loads and
+stores, (ii) operators that can trap (such as divides), and (iii)
+control flow impose hard limits on which IR-level operators can be
+merged together, and which can be hoisted or sunk in the program.
+
+For an example of a somewhat subtle condition, consider the rules that
+govern whether a load can be merged into another operator [7]:
+
+```
+    /// ...
+    /// [This function returns:]
+    /// ...
+    ///
+    /// - An instruction, given that it is effect-free or able to sink its
+    ///   effect to the current instruction being lowered, and given it has only
+    ///   one output, and if effect-ful, given that this is the only use;
+    ///
+    /// ...
+    ///
+    /// If the backend merges the effect of a side-effecting instruction, it
+    /// must call `sink_inst()`. When this is called, it indicates that the
+    /// effect has been sunk to the current scan location. The sunk
+    /// instruction's result(s) must have *no* uses remaining, because it will
+    /// not be codegen'd (it has been integrated into the current instruction).
+```
+
+[7]: https://github.com/bytecodealliance/wasmtime/blob/9419d635c68f79d12471353e28b60cc8da010dad/cranelift/codegen/src/machinst/lower.rs#L110-L125
+
+The above-cited compare-load merge bug (PR #2576) is a good example of
+a bug that can result from a non-obvious violation of these
+conditions. There are other cases where we discovered a potential
+issue during code review. We have taken care to try to design the
+`LowerCtx` API to avoid as many invariant-violations as possible
+(e.g., the `Writable<Reg>` type that encodes whether one can write to
+a given virtual register), but nevertheless, it remains an area in
+which we have to be vigilant.
+
+If we eventually express lowering logic in a DSL, then, we should
+ensure that the DSL allows us to encode such restrictions, and
+enforces them. For example, we might want the DSL to provide two
+different ways to refer to an instruction input: one when we are the
+only user, and one when we are not. Then the former might be required
+in order to sink a side-effect. Ideally, such restrictions could be
+expressed in some sort of DSL-internal type system, so that we do not
+have to hardcode lowering rules into the DSL design itself.
+
+## Requirement 6: Helps to Advance Verification Efforts
```

Comment:
Related to verification efforts: I think we should call out superoptimization as something we would like to support for our isel. We should be able to do the superoptimization offline, based on CLIF patterns that we've harvested from real world programs, and then take the learned CLIF->vcode pairs and dump them into our new DSL. We should be able to effectively fold the preopt pass (and many more peepholes!) into isel lowering.

This would give us

* better compilation throughput because we have fewer passes over the clif,
* correct by construction CLIF->vcode patterns, and
* optimal (according to some cost function) code generation for these CLIF->vcode patterns.

Reply:
That's a good point, for sure!

I would hope that the format in which we express our basic lowering patterns is general enough to support arbitrary superoptimizer-derived patterns (i.e., if this is not even technically possible then something is very wrong), so in practice it seems this boils down to, I think:

1. We should have or build a translator/bridge from a superoptimizer format into the lowering DSL (like `peepmatic-souper`)
2. We should ensure that the whole infrastructure supports the "enormous pile of complex rules" case efficiently

Reply:
Exactly, well said :)